plugins/ModelChecker/ModelChecker.py

```diff
@@ -0,0 +1,104 @@
+    def __init__(self):
+        super().__init__()
+
+        self._last_known_tool_id = None
+
+        self._controller.activeToolChanged.connect(self._onActiveToolChanged)
+
+    def checkObjects(self, nodes_to_check):
+        warning_nodes = []
+        global_container_stack = Application.getInstance().getGlobalContainerStack()
+        if global_container_stack is None:
+            return []
+        material_shrinkage = {}
+        need_check = False
+
+        # Get all shrinkage values of materials used
+        for extruder_position, extruder in global_container_stack.extruders.items():
+            shrinkage = extruder.material.getProperty("material_shrinkage_percentage", "value")
+            if shrinkage is None:
+                shrinkage = 0
+            if shrinkage > SHRINKAGE_THRESHOLD:
+                need_check = True
+            material_shrinkage[extruder_position] = shrinkage
+
+        # Check if we can bail out fast
+        if not need_check:
+            return
+
+        # Check node material shrinkage and bounding box size
+        for node in nodes_to_check:
+            node_extruder_position = node.callDecoration("getActiveExtruderPosition")
+            if material_shrinkage[node_extruder_position] > SHRINKAGE_THRESHOLD:
+                bbox = node.getBoundingBox()
+                if bbox.width >= WARNING_SIZE_XY or bbox.depth >= WARNING_SIZE_XY or bbox.height >= WARNING_SIZE_Z:
+                    warning_nodes.append(node)
+
+        return warning_nodes
+
+    def checkAllSliceableNodes(self):
+        # Add all sliceable scene nodes to check
+        scene = Application.getInstance().getController().getScene()
+        nodes_to_check = []
+        for node in DepthFirstIterator(scene.getRoot()):
+            if node.callDecoration("isSliceable"):
+                nodes_to_check.append(node)
+        return self.checkObjects(nodes_to_check)
+
+    ##  Display warning message
+    def showWarningMessage(self, warning_nodes):
+        caution_message = Message(catalog.i18nc(
+            "@info:status",
+            "Some models may not be printed optimal due to object size and material chosen [%s].\n"
+            "Tips that may be useful to improve the print quality:\n"
+            "1) Use rounded corners\n"
+            "2) Turn the fan off (only if the are no tiny details on the model)\n"
+            "3) Use a different material") % ", ".join([n.getName() for n in warning_nodes]),
+            lifetime = 0,
+            title = catalog.i18nc("@info:title", "Model Checker Warning"))
+        caution_message.show()
+
+    def showHappyMessage(self):
+        happy_message = Message(catalog.i18nc(
+            "@info:status",
+            "The Model Checker did not detect any problems with your model / print setup combination."),
+            lifetime = 5,
+            title = catalog.i18nc("@info:title", "Model Checker"))
+        happy_message.show()
+
+    def _onActiveToolChanged(self):
```

Comment:
I'm not really agreeing with that this is a tool and always visible instead of just when applicable. I'll discuss this with @LukeChen-Ultimaker though as we agreed in person.